File: python/hex_tactics/entities/skill.py
"""
Skill acquistate — port di src/entities/Skill.ts.

Vedi CLAUDE.md "Skill system" + D-046 per costi:
  total = base * (2^level - 1) / (2^specCount), arrotondato per eccesso

Modificatori base lv1 (no spec):
  -1 impedimento     100 exp
  +1 al tiro         600 exp
  +1 dado          3600 exp
  +1 dado massimo  1200 exp
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import List, Literal, Optional

from .equipment import EquipCategory, Stat

SkillModifier = Literal["-1impedimento", "+1tiro", "+1dado", "+1dadomax"]
ActionType = Literal["attaccare", "parare", "schivare", "slancio", "ricaricare"]


# Costi base lv1, no spec
SKILL_COSTS: dict[str, int] = {
    "-1impedimento": 100,
    "+1tiro": 600,
    "+1dado": 3600,
    "+1dadomax": 1200,
}


@dataclass(frozen=True)
class AcquiredSkill:
    """Una skill acquistata da un PG.

    Spec opzionali: assenti → applicazione universale.
    `level` determina sia il costo (D-046 raddoppio) che l'effetto (-N imp, +N al tiro, ecc.).
    """

    id: str
    modifier: SkillModifier
    level: int
    cost: int
    abilita: Optional[Stat] = None
    azione: Optional[ActionType] = None
    classe_oggetto: Optional[EquipCategory] = None
    oggetto_specifico: Optional[str] = None


def count_specializations(skill: AcquiredSkill) -> int:
    """0..4 specializzazioni presenti."""
    n = 0
    if skill.abilita is not None:
        n += 1
    if skill.azione is not None:
        n += 1
    if skill.classe_oggetto is not None:
        n += 1
    if skill.oggetto_specifico is not None:
        n += 1
    return n


def compute_skill_cost(modifier: SkillModifier, level: int, spec_count: int) -> int:
    """D-046: total = base * (2^level - 1) / (2^specCount), arrotondato per eccesso (Math.ceil).

    - lv1: 1× base; lv2: 3×; lv3: 7×; lv4: 15×; lvN: (2^N - 1)×
    - 0 spec: /1; 1 spec: /2; 2: /4; 3: /8; 4: /16
    """
    if level < 1:
        return 0
    base = SKILL_COSTS[modifier]
    level_multiplier = (2 ** level) - 1
    spec_divisor = 2 ** spec_count
    return math.ceil((base * level_multiplier) / spec_divisor)


def skill_key(skill: AcquiredSkill) -> str:
    """Chiave canonica per identificare 'la stessa skill' (D-046).

    Due skill con stessa chiave non possono coesistere: o sale di livello,
    o cambia spec.
    """
    return "|".join(
        [
            skill.modifier,
            skill.abilita or "",
            skill.azione or "",
            skill.classe_oggetto or "",
            skill.oggetto_specifico or "",
        ]
    )


@dataclass
class SkillSetValidation:
    valid: bool
    total_cost: int
    errors: List[str]


def validate_skill_set(skills: List[AcquiredSkill]) -> SkillSetValidation:
    """Valida un set: niente duplicati per chiave, costi totali corretti."""
    errors: List[str] = []
    seen: set[str] = set()
    total_cost = 0
    for s in skills:
        key = skill_key(s)
        if key in seen:
            errors.append(f"Skill duplicata (stessa chiave + stesso livello implicito): {key}")
        seen.add(key)
        expected = compute_skill_cost(s.modifier, s.level, count_specializations(s))
        if s.cost != expected:
            errors.append(
                f"Costo errato per {key} lv{s.level}: dichiarato {s.cost}, atteso {expected}"
            )
        total_cost += expected
    return SkillSetValidation(valid=len(errors) == 0, total_cost=total_cost, errors=errors)


@dataclass(frozen=True)
class RollContext:
    """Contesto di un tiro per matchare specializzazioni delle skill."""

    azione: ActionType
    stat: Optional[Stat] = None
    classe_oggetto: Optional[EquipCategory] = None
    oggetto_specifico: Optional[str] = None


def skill_matches_context(skill: AcquiredSkill, ctx: RollContext) -> bool:
    """Tutte le spec presenti nella skill devono matchare; quelle assenti = any."""
    if skill.azione is not None and skill.azione != ctx.azione:
        return False
    if skill.abilita is not None and skill.abilita != ctx.stat:
        return False
    if skill.classe_oggetto is not None and skill.classe_oggetto != ctx.classe_oggetto:
        return False
    if skill.oggetto_specifico is not None and skill.oggetto_specifico != ctx.oggetto_specifico:
        return False
    return True


def skill_matches_equip(skill: AcquiredSkill, equip_id: str, equip_category: EquipCategory) -> bool:
    """Match per skill `−1 impedimento` su uno specifico pezzo.

    `azione`/`abilita` ignorate (l'imp è del pezzo, non dell'azione).
    Solo `classe_oggetto` e `oggetto_specifico` contano.
    """
    if skill.classe_oggetto is not None and skill.classe_oggetto != equip_category:
        return False
    if skill.oggetto_specifico is not None and skill.oggetto_specifico != equip_id:
        return False
    return True


__all__ = [
    "SkillModifier",
    "ActionType",
    "SKILL_COSTS",
    "AcquiredSkill",
    "RollContext",
    "SkillSetValidation",
    "count_specializations",
    "compute_skill_cost",
    "skill_key",
    "validate_skill_set",
    "skill_matches_context",
    "skill_matches_equip",
]

File: src/entities/Skill.ts
/**
 * Skill acquistate dal giocatore (acquisto a punti exp).
 *
 * Vedi CLAUDE.md sezione "Skill system" per regole.
 *
 * Modificatori base (lv1):
 *   -1 impedimento     100 exp
 *   +1 al tiro         600 exp
 *   +1 dado          3600 exp   (tiri sempre 1 dado in più)
 *   +1 dado massimo  1200 exp   (tetto dei dadi tirabili +1, opzionale)
 *
 * D-046 — Costi con livelli + specializzazioni:
 *   - Lv N totale = base * (2^N - 1)
 *     Es. -1imp: lv1=100, lv2=300, lv3=700, lv4=1500, lv5=3100, lv6=6300
 *   - Ogni specializzazione DIMEZZA il costo (cumulativamente):
 *       1 spec → /2, 2 spec → /4, 3 spec → /8, 4 spec → /16
 *   - Es. -3imp [scudi] (1 spec) lv3 = 700/2 = 350
 *   - Es. +1tiro [attaccare][spade] (2 spec) lv1 = 600/4 = 150
 *
 * Vincolo unicità: non puoi acquistare la stessa skill identica più volte allo stesso livello.
 * Devi salire di livello (skill diventa lv N+1) o cambiare specializzazione.
 *
 * Specializzazioni (max 1 parola per lista, 4 liste):
 *   abilità (forza/agilità/volontà)
 *   azione (attaccare/parare/schivare/slancio/ricaricare)
 *   classe oggetto (spade/scudi/...)
 *   oggetto specifico (id arma/scudo/armatura)
 *
 * Effetto del livello: una skill di livello N equivale a "N volte" il modifier base
 * (es. -3imp toglie 3 punti dal pezzo applicabile; +2tiro [parare] = +2 fissa al parry).
 */

import { Stat, EquipCategory } from './Equipment';

/** Tipo del modificatore */
export type SkillModifier = '-1impedimento' | '+1tiro' | '+1dado' | '+1dadomax';

/** Tipo dell'azione (per specializzazione "azioni") */
export type ActionType = 'attaccare' | 'parare' | 'schivare' | 'slancio' | 'ricaricare';

/**
 * Una skill acquistata da un PG.
 * Specializzazioni opzionali: se assente, lo skill si applica universalmente.
 *
 * D-046: il livello determina sia il costo (raddoppio) sia l'effetto (-N imp, +N al tiro, ecc.).
 */
export interface AcquiredSkill {
  /** ID univoco dell'acquisto (per cumulative tracking) */
  id: string;
  modifier: SkillModifier;
  /** Livello della skill (default 1). Skill di lv N applica l'effetto N volte. */
  level: number;
  /** Specializzazione "abilità" (max 1) */
  abilita?: Stat;
  /** Specializzazione "azione" (max 1) */
  azione?: ActionType;
  /** Specializzazione "classe oggetto" (max 1, es. 'spade' / 'archi') */
  classeOggetto?: EquipCategory;
  /** Specializzazione "oggetto specifico" (max 1, id dell'item) */
  oggettoSpecifico?: string;
  /** Costo exp pagato all'acquisto (calcolato da computeSkillCost). */
  cost: number;
}

/** Costi base per modificatore (in exp), livello 1, no specializzazioni */
export const SKILL_COSTS: Record<SkillModifier, number> = {
  '-1impedimento': 100,
  '+1tiro': 600,
  '+1dado': 3600,
  '+1dadomax': 1200,
};

/** Conta quante specializzazioni ha una skill (0..4). */
export function countSpecializations(skill: Pick<AcquiredSkill, 'abilita' | 'azione' | 'classeOggetto' | 'oggettoSpecifico'>): number {
  let n = 0;
  if (skill.abilita) n++;
  if (skill.azione) n++;
  if (skill.classeOggetto) n++;
  if (skill.oggettoSpecifico) n++;
  return n;
}

/**
 * D-046: calcola il costo cumulativo di una skill al livello N con specCount specializzazioni.
 *   total = base * (2^level - 1) / (2^specCount)
 * Arrotondato per eccesso (Math.ceil) per evitare costi frazionari.
 */
export function computeSkillCost(modifier: SkillModifier, level: number, specCount: number): number {
  if (level < 1) return 0;
  const base = SKILL_COSTS[modifier];
  const levelMultiplier = Math.pow(2, level) - 1; // lv1=1, lv2=3, lv3=7, lv4=15...
  const specDivisor = Math.pow(2, specCount); // 0=1, 1=2, 2=4, 3=8, 4=16
  return Math.ceil((base * levelMultiplier) / specDivisor);
}

/**
 * D-046: chiave canonica per identificare la "stessa skill" ai fini dell'unicità.
 * Due skill con stessa chiave NON possono coesistere: o sale il livello, o cambia spec.
 */
export function skillKey(s: Pick<AcquiredSkill, 'modifier' | 'abilita' | 'azione' | 'classeOggetto' | 'oggettoSpecifico'>): string {
  return [s.modifier, s.abilita ?? '', s.azione ?? '', s.classeOggetto ?? '', s.oggettoSpecifico ?? ''].join('|');
}

/**
 * Valida un set di skill: niente duplicati per chiave, costi totali corretti.
 * @returns { valid, totalCost, errors }
 */
export function validateSkillSet(skills: AcquiredSkill[]): { valid: boolean; totalCost: number; errors: string[] } {
  const errors: string[] = [];
  const seen = new Set<string>();
  let totalCost = 0;
  for (const s of skills) {
    const key = skillKey(s);
    if (seen.has(key)) {
      errors.push(`Skill duplicata (stessa chiave + stesso livello implicito): ${key}`);
    }
    seen.add(key);
    const expected = computeSkillCost(s.modifier, s.level, countSpecializations(s));
    if (s.cost !== expected) {
      errors.push(`Costo errato per ${key} lv${s.level}: dichiarato ${s.cost}, atteso ${expected}`);
    }
    totalCost += expected;
  }
  return { valid: errors.length === 0, totalCost, errors };
}

/**
 * Contesto di un tiro, usato per matchare le specializzazioni delle skill.
 * Usato da combat.ts e ranged.ts per calcolare i modificatori applicabili.
 */
export interface RollContext {
  azione: ActionType;
  /** Stat usata nel tiro (se rilevante: attacco/parata) */
  stat?: Stat;
  /** Categoria dell'item usato (arma/scudo/armatura) */
  classeOggetto?: EquipCategory;
  /** ID specifico dell'item usato */
  oggettoSpecifico?: string;
}

/**
 * Verifica se una skill matcha il contesto di un tiro.
 * Tutte le specializzazioni presenti devono matchare; quelle assenti sono "any".
 */
export function skillMatchesContext(skill: AcquiredSkill, ctx: RollContext): boolean {
  if (skill.azione && skill.azione !== ctx.azione) return false;
  if (skill.abilita && skill.abilita !== ctx.stat) return false;
  if (skill.classeOggetto && skill.classeOggetto !== ctx.classeOggetto) return false;
  if (skill.oggettoSpecifico && skill.oggettoSpecifico !== ctx.oggettoSpecifico) return false;
  return true;
}

/**
 * Verifica se una skill `−1 impedimento` si applica a uno specifico pezzo di equipaggiamento.
 * Logica: `azione`/`abilita` sono ignorate (impedimento è del pezzo, non di un'azione).
 * Solo `classeOggetto` e `oggettoSpecifico` contano.
 *
 * Senza specializzazioni → si applica a tutti i pezzi.
 * Con `classeOggetto` → solo i pezzi di quella categoria.
 * Con `oggettoSpecifico` → solo a quel pezzo specifico.
 */
export function skillMatchesEquip(
  skill: AcquiredSkill,
  equip: { id: string; category: EquipCategory },
): boolean {
  if (skill.classeOggetto && skill.classeOggetto !== equip.category) return false;
  if (skill.oggettoSpecifico && skill.oggettoSpecifico !== equip.id) return false;
  return true;
}
